File: integer_division/integer_division.ts
// node integer_division.ts

const MIN_QUOTIENT = -Math.pow(2, 31);
const MAX_QUOTIENT = Math.pow(2, 31) - 1;

function withinRange(quotient: number) {
    if (quotient < MIN_QUOTIENT) return MIN_QUOTIENT;
    if (quotient > MAX_QUOTIENT) return MAX_QUOTIENT;
    return quotient;
}

function positive(i: number) {
    return i >= 0 ? i : ~i + 1;
}

function divide(dividend: number, divisor: number): number {
    const isNegativeQuotient = dividend < 0 !== divisor < 0;
    
    if (dividend == divisor) return 1;

    let pDividend = positive(dividend);
    const pDivisor = positive(divisor);
    let quotient = 0;
    while (pDivisor <= pDividend) {
        let jump = pDivisor;
        let multiple = 1;

        while (jump + jump <= pDividend) {
            jump += jump;
            multiple += multiple;
        }

        pDividend -= jump;
        quotient += multiple;
            
        console.log(`dividend=${pDividend}, divisor=${pDivisor}, quotient=${quotient}, jump=${jump}`)
    }

    if (isNegativeQuotient) {
        quotient = -quotient;
    }

    return withinRange(quotient);
};

type Input = Parameters<typeof convertTime>;
type Test = [Input, number];

function test() {
  const tests: Test[] = [
    [
        [10, 3],
        3,
    ],
    [
        [7, -3],
        -2,
    ],
    [
        [2147483647, -1],
        -2147483647
    ],
  ];

  for (const [input, exp] of tests) {
    const res = divide(...input);
    if (exp != res) {
      throw new Error(`Unexpected result, want=${exp}, got=${res}`);
    }
  }
}

test()

File: integer_division/integer_division.py
# uv run integers_division.py

# /// script
# requires-python = ">=3.13"
# dependencies = []
# ///

def divide(dividend, divisor):
    MAX_QUOTIENT = 2 ** 31 - 1
    MIN_QUOTIENT = -2 ** 31

    if dividend == 0: return 0
    if dividend == divisor: return 1

    is_negative = False
    if not (dividend > 0 and divisor > 0):
        is_negative = True

    quotient = 0
    dividend = abs(dividend)
    divisor = abs(divisor)

    while divisor <= dividend:
        jump = divisor
        multiple = 1

        while jump + jump <= dividend:
            jump += jump
            multiple += multiple

        dividend -= jump
        quotient += multiple

        # print(f"dividend={dividend}, divisor={divisor}, quotient={quotient}, jump={jump}")

    if is_negative:
        quotient = -quotient

    if quotient > MAX_QUOTIENT:
        return MAX_QUOTIENT

    if quotient < MIN_QUOTIENT:
        return MIN_QUOTIENT

    return quotient

def test():
    tests = [
        [
            (10, 3),
            3
        ],
        [
            (7, -3),
            -2
        ],
        [
            (10, 1),
            10
        ],
        [
            (2147483647, -1),
            -2147483647
        ],
    ]

    for input, want in tests:
        res = divide(*input)
        if want != res:
            raise RuntimeError(f"Unexpected response, want={want}, got={res}")


if __name__ == "__main__":
    test()
